pytest/lib/mocknet.py

```diff
@@ -1193,59 +1245,83 @@ def start_genesis_updater(node, script, genesis_filename_in,
+            node_pks,
+            increasing_stakes,
+            num_seats,
+            single_shard,
+            all_node_pks,
+            node_ips,
+            neard,
+        ),
+    )
 
 
-def start_genesis_update_waiter_script(done_filename):
-    return '''
-        rm /home/ubuntu/waiter.txt
-        until [ -f {done_filename} ]
-        do
-            echo 'waiting for {done_filename}' >> /home/ubuntu/waiter.txt
-            date >> /home/ubuntu/waiter.txt
-            ls {done_filename} >> /home/ubuntu/waiter.txt
-            sleep 5
-        done
-        echo 'File found' >> /home/ubuntu/waiter.txt
-    '''.format(dir=shlex.quote(PYTHON_DIR),
-               done_filename=shlex.quote(done_filename))
+def wait_genesis_updater_done(node, done_filename):
+    msg = f'Waiting for the genesis updater on {node.instance_name}'
+    logger.info(msg)
 
+    # Wait until the neard process stops running.
+    while True:
+        time.sleep(5)
+        result = node.machine.run('ps | grep neard')
+        if result.returncode != 0:
+            break
 
-def wait_genesis_updater_done(node, done_filename):
-    logger.info(f'Waiting for the genesis updater on {node.instance_name}')
-    node.machine.run('bash',
-                     input=start_genesis_update_waiter_script(done_filename))
-    logger.info(
-        f'Waiting for the genesis updater on {node.instance_name} -- done')
+    # Check if the done_filename was produced.
+    result = node.machine.run(f'ls {done_filename}')
+    if result.returncode != 0:
+        # Get the stderr of the genesis updater.
+        script = f'cat {shlex.quote(PYTHON_DIR)}/genesis_updater.err'
+        stderr = node.machine.run(script).stdout
+        error_msg = f'The {done_filename} is missing! The stderr is \n\n\n{stderr}\n'
+        raise Exception(f'{msg} - failed. {error_msg}')
 
+    logger.info(f'{msg} -- done')
 
-# Same as list_validators but assumes that the node can still be starting and expects connection errors and the data not
-# being available.
+
+# Waits until the node becomes responsive to RPC requests. It works the same as
+# list_validators but assumes that the node can still be starting and expects
+# connection errors and the data not being available.
 def wait_node_up(node):
-    logger.info(f'Waiting for node {node.instance_name} to start')
+    msg = f'Waiting for node {node.instance_name} to start'
+    logger.info(msg)
     attempt = 0
     while True:
         try:
-            node.get_validators()['result']
-            logger.info(f'Node {node.instance_name} is up')
+            response = node.get_validators()
+
+            if 'error' in response:
+                raise Exception(
+                    f'Received error from node {node.instance_name}. Error: {response["error"]}'
+                )
+            if 'result' not in response:
+                raise Exception(
+                    f'Received no result in response. Response: {response}')
```

Comment:
I think we actually expect these to happen at the beginning. When they're all computing state roots (the part that takes forever at the beginning), after the first one finishes, it will be up and looking for peers, but will be alone, with the only block in the chain so far being the genesis block. And when it's in that state, the validators rpc actually doesn't work, and you get an error saying `"UNKNOWN_EPOCH".

That said, this function will still work because of the unqualified `except Exception as e` below. Were you intending to rely on that to catch this? Honestly as part of the improvements to the mocknet spoon test I would be very in favor of getting rid of that part, since it kinda makes me nervous to just ignore any and all exceptions :/